crates/uv-resolver/src/resolution.rs

```diff
@@ -186,79 +213,136 @@ impl ResolutionGraph {
                         pins.get(package_name, version)
                             .expect("Every package should be pinned")
                             .clone()
                     };
 
-                    // Add its hashes to the index, preserving those that were already present in
-                    // the lockfile if necessary.
-                    if let Some(digests) = preferences
+                    // Extract the hashes, preserving those that were already present in the
+                    // lockfile if necessary.
+                    let hashes = if let Some(digests) = preferences
                         .match_hashes(package_name, version)
                         .filter(|digests| !digests.is_empty())
                     {
-                        hashes.insert(package_name.clone(), digests.to_vec());
+                        digests.to_vec()
                     } else if let Some(versions_response) = packages.get(package_name) {
                         if let VersionsResponse::Found(ref version_maps) = *versions_response {
-                            for version_map in version_maps {
-                                if let Some(mut digests) = version_map.hashes(version) {
+                            version_maps
+                                .iter()
+                                .find_map(|version_map| version_map.hashes(version))
+                                .map(|mut digests| {
                                     digests.sort_unstable();
-                                    hashes.insert(package_name.clone(), digests);
-                                    break;
-                                }
-                            }
+                                    digests
+                                })
+                                .unwrap_or_default()
+                        } else {
+                            vec![]
                         }
-                    }
+                    } else {
+                        vec![]
+                    };
+
+                    // Extract the metadata.
+                    let metadata = if let Some((_, metadata, _)) = editables.get(package_name) {
+                        metadata.clone()
+                    } else {
+                        let dist = PubGrubDistribution::from_registry(package_name, version);
+
+                        let response = distributions.get(&dist.version_id()).unwrap_or_else(|| {
+                            panic!(
+                                "Every package should have metadata: {:?}",
+                                dist.version_id()
+                            )
+                        });
+
+                        let MetadataResponse::Found(archive) = &*response else {
+                            panic!(
+                                "Every package should have metadata: {:?}",
+                                dist.version_id()
+                            )
+                        };
```

Comment:
I don't love these panics. (I think some like this were already in this code though.) I think this circles back to my idea of putting this data on the distribution types. Maybe that would lead to increased memory usage though? Not sure.

Reply:
We can _maybe_ put it on `ResolvedDist`? I'd have to look, but it'd be responsible to do so separately IMO since this pattern already exists here.

Reply:
Hmm, no, because we need to be able to go from compatible dist to resolved dist... At some point we need to merge the metadata with the distribution, and I guess that's here (for now, at least).